webapp/javascript/components/Heatmap/useHeatmapSelection.hook.ts

```diff
@@ -0,0 +1,296 @@
+        y: HEATMAP_HEIGHT - (cellMatrixCoordinate[1] + 1) * cellH,
+      };
+
+      fetchProfile(
+        startCoords.x,
+        endCoords.x,
+        startCoords.y,
+        endCoords.y,
+        startCoords.y === HEATMAP_HEIGHT
+      );
+    }
+  };
+
+  const startDrawing = (e: MouseEvent) => {
+    window.addEventListener('mousemove', handleDrawingEvent);
+    window.addEventListener('mouseup', endDrawing);
+
+    const canvas = canvasRef.current as HTMLCanvasElement;
+    const { left, top } = canvas.getBoundingClientRect();
+    resetSelection();
+
+    startCoords = { x: e.clientX - left, y: e.clientY - top };
+  };
+
+  const endDrawing = (e: MouseEvent) => {
+    if (startCoords) {
+      const canvas = canvasRef.current as HTMLCanvasElement;
+      const { left, top, width, height } = canvas.getBoundingClientRect();
+      setHasSelectedArea(true);
+      clearRect(canvas);
+
+      const xCursorPosition = e.clientX - left;
+      const yCursorPosition = e.clientY - top;
+      let xEnd;
+      let yEnd;
+
+      if (xCursorPosition < 0) {
+        xEnd = 0;
+      } else if (xCursorPosition > width) {
+        xEnd = width;
+      } else {
+        xEnd = xCursorPosition;
+      }
+
+      if (yCursorPosition < 0) {
+        yEnd = 0;
+      } else if (yCursorPosition > height) {
+        yEnd = parseInt(height.toFixed(0), 10);
+      } else {
+        yEnd = yCursorPosition;
+      }
+
+      endCoords = { x: xEnd, y: yEnd };
+      const isClickEvent = startCoords.x === xEnd && startCoords.y === yEnd;
+
+      if (isClickEvent) {
+        handleCellClick(xEnd, yEnd);
+      } else {
+        fetchProfile(startCoords.x, xEnd, startCoords.y, yEnd);
+      }
+
+      window.removeEventListener('mousemove', handleDrawingEvent);
+      window.removeEventListener('mouseup', endDrawing);
+
+      const selectedAreaW = endCoords.x - startCoords.x;
+      if (selectedAreaW) {
+        selectedAreaToHeatmapRatio = Math.abs(width / selectedAreaW);
+      } else {
+        selectedAreaToHeatmapRatio = 1;
+      }
+    }
+  };
+
+  const handleDrawingEvent = (e: MouseEvent) => {
+    const canvas = canvasRef.current as HTMLCanvasElement;
+
+    if (canvas && startCoords) {
+      const { left, top } = canvas.getBoundingClientRect();
+
+      /**
+       * Cursor coordinates inside canvas
+       * @cursorXCoordinate - e.clientX - left
+       * @cursorYCoordinate - e.clientY - top
+       */
+      const width = e.clientX - left - startCoords.x;
+      const h = e.clientY - top - startCoords.y;
+
+      drawRect(canvas, startCoords.x, startCoords.y, width, h);
+    }
+  };
+
+  useEffect(() => {
+    if (from && until && query) {
+      const fetchData = dispatch(
+        fetchHeatmapSingleView({
+          query,
+          from,
+          until,
+          ...DEFAULT_HEATMAP_PARAMS,
+        })
+      );
+      return () => fetchData.abort('cancel');
+    }
+    return undefined;
+  }, [from, until, query]);
```

Comment:
I think this is waaaay too much functionality for this single hook. Let's move data fetching to the component/another hook.